### ISSUE TITLE
ale_lint_on_insert_leave=1 isn't working as expected

### ISSUE THREAD

**Issue by sbromberger** [dense-analysis/ale#2964]
## Information

**VIM version**
```
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Dec 27 2019 15:11:16)
Included patches: 1-50
```

Operating System: <!-- Describe your operating system version. -->
`Linux hydrogen 5.3.0-24-generic #26-Ubuntu SMP Thu Nov 14 01:33:18 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`

## What went wrong
Setting `ale_lint_on_insert_leave=1` does not run the linter after using `<ESC>` to leave insert mode. 
## Reproducing the bug

I inserted the following go code:
```
func dothis() {
}
```
and then hit `<ESC>`, and expected the linter to warn me of dead code. It did not warn me until I explicitly saved the file (`:w`).

### :ALEInfo

```
Current Filetype: go
Available Linters: ['bingo', 'gobuild', 'gofmt', 'golangci-lint', 'golint', 'gometalinter', 'gopls', 'gosimple', 'gotype', 'govet', 'golangse
rver', 'staticcheck']
   Linter Aliases:
'gobuild' -> ['go build']
'govet' -> ['go vet']
  Enabled Linters: ['golangci-lint', 'gopls']
 Suggested Fixers:
  'gofmt' - Fix Go files with go fmt.
  'goimports' - Fix Go files imports with goimports.
  'remove_trailing_lines' - Remove all blank lines at the end of a file.
  'trim_whitespace' - Remove all trailing whitespace characters at the end of every line.
 Linter Variables:
let g:ale_go_go_executable = 'go'
let g:ale_go_golangci_lint_executable = 'golangci-lint'
let g:ale_go_golangci_lint_options = '--enable golint'
let g:ale_go_golangci_lint_package = 1
let g:ale_go_gopls_executable = 'gopls'
let g:ale_go_gopls_options = '--mode stdio'
let g:ale_go_langserver_executable = 'go-langserver'
let g:ale_go_langserver_options = ''
 Global Variables:
let g:ale_cache_executable_check_failures = v:null

let g:ale_change_sign_column_color = 0
let g:ale_command_wrapper = ''
let g:ale_completion_delay = v:null
let g:ale_completion_enabled = 0
let g:ale_completion_max_suggestions = v:null
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%code: %%s'
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {'go': ['goimports'], 'python': ['autopep8']}
let g:ale_history_enabled = 1
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 20
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_insert_leave = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'normal'
let g:ale_linter_aliases = {}

let g:ale_linters = {'go': ['gopls', 'golangci-lint'], 'md': ['mdl'], 'python': ['pyls'], 'julia': ['languageserver']}
let g:ale_linters_explicit = 1
let g:ale_list_vertical = 0
let g:ale_list_window_size = 10
let g:ale_loclist_msg_format = '%code: %%s'
let g:ale_lsp_root = {}
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = -1
let g:ale_maximum_file_size = v:null
let g:ale_open_list = 0
let g:ale_pattern_options = v:null
let g:ale_pattern_options_enabled = v:null
let g:ale_set_balloons = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 1
let g:ale_sign_error = '>>'
let g:ale_sign_info = '--'
let g:ale_sign_offset = 1000000
let g:ale_sign_style_error = '>>'
let g:ale_sign_style_warning = '--'

let g:ale_sign_highlight_linenrs = 0
let g:ale_statusline_format = v:null
let g:ale_type_map = {'golangci-lint': {'E': 'WS', 'ES': 'WS'}}
let g:ale_use_global_executables = v:null
let g:ale_virtualtext_cursor = 0
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
//// sensitive command history elided - nothing useful there.
```

vimrc is simpler:
```
set nocompatible                                                                                                                             
set encoding=UTF-8                                                                                                                           
set fileencoding=utf8                                                                                                                        
set redrawtime=10000                                                                                                                         
scriptencoding utf-8                                                                                                                         
set t_Co=256                                                                                                                                 
set timeoutlen=1000 ttimeoutlen=10 " sets delay for escape char                                                                              
" set clipboard=unnamed " copy from elsewhere, paste using 'p'                                                                               
                                                                                                                                             
set number              " line numbering                                                                                                     
set ruler                                                                                                                                    
set fillchars+=vert:\│  " vertical window separator                                                                                          
set noshowmode          " disable show mode since liteline has it already                                                                    
set history=1000        " save 1000 ex commands                                                                                              
set hlsearch            " highlight search results                                                                                           
set incsearch           " highlight search results incrementally                                                                             
set cursorline          " highlight line that cursor is on                                                                                   
nnoremap <C-L> :noh<CR><C-L>                                                                                                                 
" activate visual mode in normal mode                                                                                                        
nmap <S-Up> Vk                                                                                                                               
nmap <S-Down> Vj                                                                                                                             
" these are mapped in visual mode                              
vmap <S-Up> k                                                                                                                        [5/1950]
vmap <S-Down> j                                                                                                                              
                                                                                                                                             
let mapleader=","       " set leader to comma                                                                                                
let g:ale_completion_enabled = 0                                                                                                             
let g:ale_lint_on_text_changed='normal'                                                                                                      
let g:ale_lint_on_insert_leave=1
let g:ale_lint_delay=20
call plug#begin('~/.vim/plugged')
Plug 'dense-analysis/ale'                                       " this provides linting and error checking.
call plug#end()
syntax on
"{{{ ALE configuration
  let g:ale_sign_column_always = 1      "Leave enough space in the gutter for errors and warnings
  let g:ale_linters_explicit = 1
  let g:ale_linters = { 'go': ['gopls', 'golangci-lint'], 'python':['pyls'], 'md':['mdl'], 'julia':['languageserver'] }
  " let g:ale_go_gopls_options = '-v -rpc.trace -logfile /tmp/plslog.txt'
  let g:ale_go_golangci_lint_options = '--enable golint'
  let ale_go_golangci_lint_package = 1
  let g:ale_fixers = { 'go': ['goimports'], 'python': ['autopep8'] }
  let g:ale_type_map = {'golangci-lint': {'ES': 'WS', 'E': 'WS'}}
nmap <silent> <C-k> <Plug>(ale_previous_wrap)
  nmap <silent> <C-j> <Plug>(ale_next_wrap)
  nnoremap <silent> K :ALEHover<cr>
"}}}
```

**Comment by mrbeardad** [comment]
I have the similar question: When I press <esc> to leave Insert Mod, only gcc and clang will work, clangcheck only work when write the file.
`let g:ale_linters = {`
`      \   'cpp': ['gcc', 'cppcheck', 'clang','clangcheck'],`
`      \   'c': ['gcc', 'cppcheck', 'clang'],`
`      \   'python': ['pylint'],`
`      \   'sh': ['shell'],`
`\}`
`let g:ale_linters_explicit = 1`
`let g:ale_completion_delay = 500`
`let g:ale_echo_delay = 20`
`let g:ale_lint_delay = 500`
`let g:ale_set_highlights = 0`
`let g:ale_echo_msg_format = '[%linter%] %code: %%s'`
`let g:ale_lint_on_text_changed = 'normal'`
`let g:ale_lint_on_insert_leave = 1`
`let g:ale_lint_on_enter = 0`
`let g:ale_c_gcc_options = '-Wall -O2 -std=c11'`
`let g:ale_c_clang_options = '-Wall -O2 -std=c11'`
`let g:ale_c_cppcheck_options = ''`
`let g:ale_cpp_gcc_options = '-Wall -O2 -std=c++17'`
`let g:ale_cpp_clang_options = '-Wall -O2 -std=c++17'`
`let g:ale_cpp_cppcheck_options = ''`
`let g:ale_cpp_clangcheck_options = ''`
`let g:ale_sign_error = '✗'`
`let g:ale_sign_warning = '⚡'`
`hi! clear SpellBad`
`hi! clear SpellCap`
`hi! clear SpellRare`
`hi! SpellBad gui=undercurl guisp=red`
`hi! SpellCap gui=undercurl guisp=blue`
`hi! SpellRare gui=undercurl guisp=magenta`
`nmap <silent> jb <Plug>(ale_previous_wrap)`
`nmap <silent> jn <Plug>(ale_next_wrap)`

**Comment by mrbeardad** [comment]
More specifically, as long as I `let g:ale_cpp_gcc_options = ' any options '`, ale will not work correctly: some error messages will not display by ale

**Comment by w0rp** [comment]
`go build` and `go vet` are linters that are only capable of running against files written to disk, which is true of most Go tools. There's no issue here.
